### PR TITLE
fix: attach to v8 debugger session only when having active breakpoints

### LINKED ISSUE #811: Potential Memory Leak
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

Please run down the following list and make sure you've tried the usual "quick fixes":

  - Search the issues already opened: https://github.com/googleapis/cloud-debug-nodejs/issues
  - Search the issues on our "catch-all" repository: https://github.com/googleapis/google-cloud-node
  - Search StackOverflow: http://stackoverflow.com/questions/tagged/google-cloud-platform+node.js

If you are still having issues, please be sure to include as much information as possible:

#### Environment details

  - OS: Google App Engine
  - Node.js version: 12.14.x
  - npm version: 6.13
  - `@google-cloud/debug-agent` version: 4.2.1

#### Steps to reproduce
We found that the package cause a memory leak. This image shows our application memory with the package and without it. 

image

We are running on Google App Engine with 4 nodes. Any help is appreciated. 

Thanks!

### PR DESCRIPTION
**Pull request by Louis-Ye**
* This PR also refactos the inspector-debugapi class so that the V8
related code is moved to v8inspector class.

Fixes #811

### CODE REVIEW

**Review by JustinBeckwith — APPROVED** [review]

File: src/agent/v8/inspector-debugapi.ts [hunk 1]
```diff
@@ -108,36 +83,28 @@ export class InspectorDebugApi implements debugapi.DebugApi {
...
-      this.scriptMapper[script.params.scriptId] = script.params;
-    });
-    session.post('Debugger.enable');
-    session.post('Debugger.setBreakpointsActive', {active: true});
-    session.on('Debugger.paused', message => {
-      try {
-        this.handleDebugPausedEvent(message.params);
-      } catch (error) {
-        this.logger.error(error);
+    this.inspector = new V8Inspector(
```
Comment (mctavish) [inline src/agent/v8/inspector-debugapi.ts:86]: useWellFormattedUrl is in here twice.  One of them is mislabeled.

File: src/agent/config.ts [hunk 2]
```diff
@@ -359,10 +359,10 @@ export interface ResolvedDebugAgentConfig extends GoogleAuthOptions {
   apiUrl?: string;
 
   /**
-   * Number of times of the V8 breakpoint hits events before resetting the
-   * breakpoints. This is to release the memory usage held by V8 engine for each
-   * breakpoint hit to prevent the memory leak. The default value is specified
-   * in defaultConfig.
+   * Number of times of the V8 pause events (could be breakpoint hits) before
```
Comment (mctavish) [inline src/agent/config.ts:362]: Nit: s/pause/pauses ?

File: src/agent/v8/v8inspector.ts [hunk 3]
```diff
@@ -92,4 +142,63 @@ export class V8Inspector {
...
+    this.storeSetBreakpointParams = {};
+    this.numPausedBeforeReset = 0;
+  }
+
+  /**
+   * Resets the debugging session when the threshold. This is primarily for
+   * cleaning the memory usage hold by V8 debugger when hitting the V8
+   * breakpoints too many times.
+   */
+  private resetV8DebuggerIfMeetThreshold() {
```
Comment (mctavish) [inline src/agent/v8/v8inspector.ts:184]: Nit: IfThresholdMet reads a bit better.

File: src/agent/v8/v8inspector.ts [hunk 4]
```diff
@@ -92,4 +142,63 @@ export class V8Inspector {
...
+    session.post('Debugger.setBreakpointsActive', {active: true});
+    session.on('Debugger.paused', message => {
+      this.onPaused(message.params);
+      this.resetV8DebuggerIfMeetThreshold();
+    });
+
+    this.session = session;
+  }
+
+  /** Detaches from the V8 debugger. */
```
Comment (mctavish) [inline src/agent/v8/v8inspector.ts:167]: Please add a note that it purges breakpoint state (ie. detach + attach is lossy)

**Review by mctavish — CHANGES_REQUESTED** [review]
A few nits; otherwise looks good.

File: src/agent/v8/inspector-debugapi.ts (hunk 1 above)
Comment (Louis-Ye) [reply src/agent/v8/inspector-debugapi.ts:90]: Done.

File: src/agent/config.ts (hunk 2 above)
Comment (Louis-Ye) [reply src/agent/config.ts:362]: Done.

File: src/agent/v8/v8inspector.ts (hunk 3 above)
Comment (Louis-Ye) [reply src/agent/v8/v8inspector.ts:184]: Done.

File: src/agent/v8/v8inspector.ts (hunk 4 above)
Comment (Louis-Ye) [reply src/agent/v8/v8inspector.ts:167]: Done.

**Review by mctavish — APPROVED** [review]

**Pull request merged**